src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/handler/LagPortUpdateHandler.java

```diff
@@ -0,0 +1,187 @@
+                formatLagPortReference(), formatTargetPorts(goal.getTargetPorts()));
+        carrier.sendCommandToSpeaker(request, cookie);
+        pendingSpeakerRequests.add(cookie);
+    }
+
+    /**
+     * Handle GRPC response.
+     */
+    public void dispatchGrpcResponse(CreateOrUpdateLogicalPortResponse response, MessageCookie cookie) {
+        if (! pendingSpeakerRequests.remove(cookie)) {
+            logUnwantedResponse(response);
+            return;
+        }
+
+        log.info("{} have been updated", formatLagPortReference());
+        carrier.response(requestKey, new LagPortResponse(goal.getLogicalPortNumber(), goal.getTargetPorts()));
+    }
+
+    /**
+     * Handle GRPC error response.
+     */
+    public void dispatchGrpcResponse(ErrorData response, MessageCookie cookie) {
+        if (!pendingSpeakerRequests.remove(cookie)) {
+            logUnwantedResponse(response);
+            return;
+        }
+
+        log.error("Unable to update {}: {}", formatLagPortReference(), response);
+
+        // TODO(surabujin): should we retry update attempt?
+
+        fail(response.getErrorType(), response.getErrorMessage());
+    }
+
+    /**
+     * Handle timeout event.
+     */
+    public void timeout() {
+        fail(ErrorType.OPERATION_TIMED_OUT, "Timeout communication switch via GRPC");
+
+        pendingSpeakerRequests.clear(); // force handle completion
+    }
+
+    public boolean isCompleted() {
+        return pendingSpeakerRequests.isEmpty();
+    }
+
+    private void fail(ErrorType type, String description) {
+        String errorMessage;
+        if (rollback()) {
+            errorMessage = String.format("Unable to update %s", formatLagPortReference());
+        } else {
+            errorMessage = String.format(
+                    "Unable to update %s, also DB data rollback have failed, use switch validate/sync to restore "
+                            + "system's consistent state",
+                    formatLagPortReference());
+        }
+
+        carrier.errorResponse(requestKey, type, errorMessage, description);
+    }
+
+    private boolean rollback() {
+        try {
+            operationService.updateLagPort(goal.getSwitchId(), goal.getLogicalPortNumber(), rollbackTargets);
+        } catch (SwitchManagerException e) {
+            log.error("Unable to rollback DB update for {}: {}", formatLagPortReference(), e.getMessage());
+            return false;
+        }
+        return true;
+    }
+
+    private void logUnwantedResponse(MessageData payload) {
+        log.info("Got unwanted/outdated GRPC response: {}", payload);
+    }
+
+    private MessageCookie newMessageCookie() {
+        for (int i = 0; i < COOKIE_GENERATION_ATTEMPTS_LIMIT; i++) {
+            MessageCookie attempt = new MessageCookie(cookieValueGenerator.generate().toString());
+            if (pendingSpeakerRequests.contains(attempt)) {
+                continue;
+            }
+
+            return attempt;
+        }
```

Comment:
I guess uuid generate attempts are redundant. Probability is the same as to generate two equal correlation-id for requests.

Reply:
It costs one simple enouth loop. I feels uncomphortable because of possible collistion, and this is the way how to get dial with it.